### ISSUE TITLE
Allow field data to be stored on disk for analyzed string fields

### ISSUE THREAD

**Issue by jmacdonald92** [elastic/elasticsearch#13400]
Hello, sorry if there are already issues related to this topic. 

I am wondering if it would 1) be possible 2) be a good idea to allow the fielddata data structure to be stored on disk and loaded into the file system cache when needed, similar to doc_values. I know that allowing doc_values storage for analyzed string fields would not work for performance reasons (way too many values per document, for one). How I imagine this would work is: fielddata is build as it currently is and then serialized and written to disk; then when it is needed it is loaded into the file system cache. The benefit here is decreased memory use while (hopefully) still avoiding the cold query latency spike that occurs when you aggregate over an index without eager loaded field data for the first time (I assume that is primarily the cost of un-inverting a large index and building the fielddata structure, and that paging in that structure from file is faster). 

I need this feature because I use Elasticsearch for several aggregations on analyzed text (including a terms aggregation) for over a billion records across hundreds of daily indices. Unfortunately, this means that I need to eager load fielddata for all of these indices to prevent 20-200 second query latency on historical queries (which would also likely evict more recent field data and add GC latency). Currently we need around 90gb of RAM just for field data on one replica of our data, which is fairly expensive, and we think it would reduce costs of our Elasticsearch deployment significantly if we could move some of that storage requirement to disk.

Thanks!

**Comment by jpountz** [comment]
What kind of text do you analyze? If this is simple text that you just happen to normalize (lowercasing, removal of accents, etc.), this other issue should allow you to enable doc values on it: #12394

**Comment by jmacdonald92** [comment]
Sorry I was not clear about that. The primary use of the field is terms aggregation to find the top tokens in a slice of the data. We analyze the data using the standard analyzer. So basically, the only type of data that will not have doc_values as the default post #12394. 

I know that allowing doc_values for tokenized string data would have terrible performance. I was wondering if it would be possible to store the current fielddata structure for these string fields on disk and keep it updated/page it in to the FS cache like doc_values. From what I understand, the problems with enabling doc values stem from the data structures used, but it appears that whichever data structure is backing the fielddata implementations doesn't have this issue.

**Comment by jpountz** [comment]
OK, so this is actual full text content and wouldn't be covered by #12394. 

Indeed doc values are good to store string fields when documents have at most a couple values. Full text makes it harder because there can be a lot of variance in the number of terms that each document has, which is hard to store efficiently.

**Comment by rmuir** [comment]
Fielddata is just as bad, it just blows up memory faster. Again, frequencies are lost so its just a bad idea for tokenized text.

**Comment by jmacdonald92** [comment]
I am either fundamentally miss-understanding how fielddata works in the context of aggregations or @rmuir and I are not talking about the same thing/use case. 

As far as I understand, when Elasticsearch performs an aggregation it has to un-invert the stored index first. By default it does this with field data, but this process causes a latency spike if the field data isn't in the cache, since the process of constructing it (and potentially garbage collecting whatever was in the cache previously) is computationally expensive. The alternative is to index the field with doc_values = true, which constructs an inverted index using a different data structure and saves it to disk. Then when that doc_values data is needed the file on disk is page in to the file system cache outside the JVM heap. A cache miss here is much less problematic, since there is no need to garbage collect and the data structure is pre-built, since the only cost is reading from disk (usually an SSD). 

Given that understanding (which is likely wrong), I am confused about the statement that fielddata is bad for tokenized text. As far as I understand, fielddata is currently the only un-inverted form Elasticsearch supports for tokenized text. I absolutely agree that is takes a lot of memory, however, I have never encountered an issue with performance. Also, I don't know why frequencies being lost is problematic, since as far as I know the terms aggregation never supported terms frequency anyway (and if I am wrong and it does, I am not sure how).

Anyway, what I am really looking for is a way to replicate the job the fielddata currently performs for tokenized text without the requirement of having it constantly loaded into memory or face terrible performance.

**Comment by rmuir** [comment]
Thats the issue though, frequencies are important for tokenized text. I think somehow for whatever reason, your use case is ok with this, but generally speaking it is the wrong datastructure. I dont think we should try to provide enhancements around fielddata or docvalues support for that reason alone.

**Comment by jmacdonald92** [comment]
I agree 100% that frequencies are important for tokenized text and that for that reason (among others) fielddata and docvalues are not a great format for storing and using an un-inverted index of tokenized text. What has me confused is that, as far as I know, fielddata is the only format Elasticsearch currently uses for that purpose. 

At the end of the day, the thing I really care about is reducing the memory requirement to run Elasticsearch. As far as I know, that means I need some way to store the data needed for aggregations in a way that does not take a long time to retrieve, or have to be held in memory constantly. As far as I know, we currently have this option for every other field type in dovvalues. So what I am really looking for is not specific to either fielddata or docvalues. 

That said, (I think) Elasticsearch currently uses fielddata (only) to fulfill this purpose for tokenized string fields. So my thought was saving the fielddata structure to disk and treating it like a cache backed by files on disk. I have no idea if that is the best solution, or if it is even possible but I wanted to start a discussion about moving (or having the option to move) tokenized string field storage to disk in some way if possible.

**Comment by jmacdonald92** [comment]
I wanted to update that I have been doing some digging into the way fielddata is generated, stored and used in Elasticsearch, and I am starting to realize that adding an FS backed cache not be possible.

I may be wrong about this, but as far as I could tell fielddata is comprised of some sort of tree of lucene data structures containing that allows you to map global ordinals to segment specific ordinals (which is obviously a gross oversimplification). The potential problem I am seeing is that these data structures are likely not serializable. I am not sure if there is a better or more possible way to store fielddata, but as @rmuir  said, fielddata is probably not the best way to accomplish the goal anyway. 

Anyway, I am still interested in some solution that would allow us stop keeping fielddata for tokenized strings in memory constantly without negatively impacting performance. 

Thanks

**Comment by clintongormley** [comment]
@jmacdonald92 There appear to be two different use cases for doc values (or some other structure) with analyzed string fields. One is just the normalization case (eg lower-casing all terms), which is covered by https://github.com/elastic/elasticsearch/issues/10061.

The other use case is harder to pin down.  Suggested purposes include tag clouds, aggregated term statistics for naive bayes training, feature selection, and query expansion.

The problem is coming up with an efficient data structure which supports these cases.  Some require term statistics, and some don't. But all of them look pretty heavy. For the significant terms agg, we provide the sampler aggregation to reduce the load (see https://www.elastic.co/guide/en/elasticsearch/reference/2.0/search-aggregations-bucket-sampler-aggregation.html), but more details about real use cases would be useful.

**Comment by colings86** [comment]
We discussed this internally and we are not looking to implement this in the foreseeable future. Since this issue was opened we have dedicated keyword fields with normalisers supported and field data on text fields is largely no longer necessary. For this reason we are going to close this issue for now.

If anyone coming across this issue has a specific use case that they require field data on text fields for that is not covered by existing functionality please open a new issue with details on the exact use case and feature request and we can look into the best way to support that use case (hopefully without needing to use field data on the text field)
